### PR TITLE
k*c instance weighting.

### PR DESCRIPTION
**Pull request by olliethomas**

### CODE REVIEW

File: bolts/losses/loss.py [hunk 1]
```diff
@@ -1,13 +1,44 @@
...
 
 class CrossEntropy(nn.CrossEntropyLoss):
-    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+    def __init__(
+        self,
+        weight: Optional[Tensor] = None,
+        size_average=None,
+        ignore_index: int = -100,
+        reduce=None,
+        reduction: str = 'mean',
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:18]: You could be a good boy and make this a literal

File: bolts/losses/loss.py [hunk 2]
```diff
@@ -1,13 +1,44 @@
...
 
 __all__ = ["CrossEntropy", "OnlineReweightingLoss"]
 
 
 class CrossEntropy(nn.CrossEntropyLoss):
-    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+    def __init__(
+        self,
+        weight: Optional[Tensor] = None,
+        size_average=None,
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:15]:
```suggestion
```
this is deprecated

File: bolts/losses/loss.py [hunk 3]
```diff
@@ -1,13 +1,44 @@
...
+        ignore_index: int = -100,
+        reduce=None,
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:17]:
```suggestion
```
This is also deprecated

File: bolts/losses/loss.py [hunk 4]
```diff
@@ -1,13 +1,44 @@
...
+        reduction: str = 'mean',
+    ) -> None:
+        super().__init__(weight=weight, size_average=size_average, reduce=reduce, reduction="none")
+        self.ignore_index = ignore_index
+        self._reduction_str = reduction
+
+    def forward(self, input: Tensor, target: Tensor, weight: Optional[Tensor] = None) -> Tensor:
         _target = target.view(-1).long()
-        return super().forward(input, _target)
+        _weight = weight.view(-1) if weight is not None else torch.ones_like(_target)
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:26]: why not just multiply conditionally?

File: bolts/losses/loss.py [hunk 5]
```diff
@@ -1,13 +1,44 @@
...
+        losses = F.cross_entropy(
+            input,
+            _target,
+            weight=self.weight,
+            ignore_index=self.ignore_index,
+            reduction=self.reduction,
+        )
+        losses *= _weight
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:34]:
```suggestion
        if weight is not None:
            _weight = weight.flatten()
            losses *= _weight
```

File: bolts/losses/loss.py [hunk 6]
```diff
@@ -1,13 +1,44 @@
...
+import torch
 from torch import Tensor, nn
+import torch.nn.functional as F
 from torch.nn.modules.loss import _Loss
+ main
 
 __all__ = ["CrossEntropy", "OnlineReweightingLoss"]
 
 
 class CrossEntropy(nn.CrossEntropyLoss):
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:12]:
```suggestion
class CrossEntropyLossIW(nn.CrossEntropyLoss):
```
